coderd/tailnet_test.go

```diff
@@ -95,6 +95,32 @@ func TestServerTailnet_ReverseProxy(t *testing.T) {
 		assert.Equal(t, http.StatusOK, res.StatusCode)
 	})
 
+	t.Run("HostRewrite", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		agentID, _, serverTailnet := setupAgent(t, nil)
+
+		u, err := url.Parse(fmt.Sprintf("http://127.0.0.1:%d", codersdk.WorkspaceAgentHTTPAPIServerPort))
+		require.NoError(t, err)
+
+		rp, release, err := serverTailnet.ReverseProxy(u, u, agentID)
+		require.NoError(t, err)
+		defer release()
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+		require.NoError(t, err)
+
+		// Ensure the reverse proxy director rewrites the url host to the agent's IP.
+		rp.Director(req)
+		assert.Equal(t,
+			fmt.Sprintf("[%s]:%d", tailnet.IPFromUUID(agentID).String(), codersdk.WorkspaceAgentHTTPAPIServerPort),
+			req.URL.Host,
+		)
+	})
```

Comment:
This is a good test, but I'd also like to see an integration test where we create 2 workspaces with applications running on the same port, then use the server tailnet to dial them.  A lack of a test like that allowed the issue to go unnoticed. 

It would be good to verify that 

a) the HTTP requests hit the correct workspaces
b) the HTTP transport does cache the connections (e.g. issuing two requests to each workspace and verify we Dial once - you can do this with a wrapped Listener that you bind the HTTP servers to).